File: LeetCode_878/Main.java
package LeetCode_878;
class Solution {
    public int nthMagicalNumber(int n, int a, int b) {
        final int MOD = 1_000_000_007;
        long A = a, B = b;
        long LCM = lcm(A, B);

        long lo = 1, hi = (long) 1e18, ans = 1;

        while (lo <= hi) {
            long mid = lo + (hi - lo) / 2;
            long count = mid / A + mid / B - mid / LCM;

            if (count >= n) {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }

        return (int) (ans % MOD);
    }

    private long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }
}

public class Main {
    public static void main(String[] args) {
        
    }
}

File: LeetCode_878/Main.js
function gcd(x, y) {
    while (y !== 0n) {
        let temp = y;
        y = x % y;
        x = temp;
    }
    return x;
}

function lcm(x, y) {
    return x / gcd(x, y) * y;
}

function isGood(mid, a, b, n) {
    let ab = lcm(a, b);
    let count = mid / a + mid / b - mid / ab;
    return count >= n;
}

/**
 * @param {number} n
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
var nthMagicalNumber = function(n, a, b) {
    const MOD = 1000000007n;

    // Convert inputs to BigInt
    n = BigInt(n);
    a = BigInt(a);
    b = BigInt(b);

    let lo = 1n, hi = 1000000000000000000n; // 1e18 as BigInt
    let ans = 1n;

    while (lo <= hi) {
        let mid = lo + (hi - lo) / 2n;
        if (isGood(mid, a, b, n)) {
            ans = mid;
            hi = mid - 1n;
        } else {
            lo = mid + 1n;
        }
    }

    return Number(ans % MOD);  // Convert back to Number for final result
};
